File: src/main/java/cc/mirukuneko/realtrainmodrenewed/client/ClientHooksClient.java
package cc.mirukuneko.realtrainmodrenewed.client;

import cc.mirukuneko.realtrainmodrenewed.client.screen.MarkerConfigScreen;
import cc.mirukuneko.realtrainmodrenewed.client.screen.SpeakerScreen;
import cc.mirukuneko.realtrainmodrenewed.blockentity.InstalledObjectBlockEntity;
import cc.mirukuneko.realtrainmodrenewed.client.screen.ScriptBlockScreen;
import cc.mirukuneko.realtrainmodrenewed.client.screen.SignalChangerScreen;
import cc.mirukuneko.realtrainmodrenewed.client.screen.SignalReceiverScreen;
import cc.mirukuneko.realtrainmodrenewed.client.screen.SignalValueScreen;
import cc.mirukuneko.realtrainmodrenewed.client.screen.TrainDetectorScreen;
import cc.mirukuneko.realtrainmodrenewed.client.sound.CrossingGateSoundManager;
import cc.mirukuneko.realtrainmodrenewed.installedobject.InstalledObjectCategory;
import cc.mirukuneko.realtrainmodrenewed.item.TrainItem;
import net.minecraft.client.Minecraft;
import net.minecraft.core.BlockPos;
import net.minecraft.network.chat.Component;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.api.distmarker.OnlyIn;

@OnlyIn(Dist.CLIENT)
public final class ClientHooksClient {
    private ClientHooksClient() {
    }

    public static void openRailSelectScreen(Player player, ItemStack stack) {
        ClientItemHelper.openRailSelectScreen(player, stack);
    }

    public static void openTrainSelectScreen(Player player, ItemStack stack, TrainItem.Category category) {
        ClientItemHelper.openTrainSelectScreen(player, stack, category);
    }

    public static void openTrainSelectScreen(Player player, ItemStack stack) {
        ClientItemHelper.openTrainSelectScreen(player, stack);
    }

    public static void openVehicleFormationScreen(ItemStack stack) {
        ClientItemHelper.openVehicleFormationScreen(stack);
    }

    public static void openCarSelectScreen(Player player, ItemStack stack) {
        ClientItemHelper.openCarSelectScreen(player, stack);
    }

    public static void openInstalledObjectSelectScreen(Player player, ItemStack stack, InstalledObjectCategory category) {
        ClientItemHelper.openInstalledObjectSelectScreen(player, stack, category);
    }

    public static void openSignalChangerScreen(BlockPos pos) {
        Minecraft.getInstance().setScreen(new SignalChangerScreen(pos));
    }

    public static void openSignalReceiverScreen(BlockPos pos) {
        Minecraft.getInstance().setScreen(new SignalReceiverScreen(pos));
    }

    public static void openSignalValueScreen(BlockPos pos) {
        Minecraft.getInstance().setScreen(new SignalValueScreen(pos));
    }

    public static void openTrainDetectorScreen(BlockPos pos) {
        Minecraft.getInstance().setScreen(new TrainDetectorScreen(pos));
    }

    public static void openMarkerConfigScreen(BlockPos pos) {
        Minecraft.getInstance().setScreen(
            new MarkerConfigScreen(pos));
    }

    public static void openSpeakerScreen(BlockPos pos) {
        Minecraft.getInstance().setScreen(
            new SpeakerScreen(pos));
    }

    public static void openScriptBlockScreen(BlockPos pos) {
        Minecraft.getInstance().setScreen(new ScriptBlockScreen(pos));
    }

    public static void stopCrossingGateSound(Level level, BlockPos pos) {
        CrossingGateSoundManager.stop(level, pos);
    }

    public static void tickCrossingGateSound(InstalledObjectBlockEntity blockEntity) {
        CrossingGateSoundManager.tick(blockEntity);
    }

    public static void showScriptErrorMessage(String message) {
        Minecraft minecraft = Minecraft.getInstance();
        if (minecraft.player == null || message == null || message.isBlank()) {
            return;
        }
        minecraft.player.sendSystemMessage(Component.literal("[RTMU Script] " + message));
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/client/ClientHooksClient.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.client

import cc.mirukuneko.realtrainmodrenewed.blockentity.InstalledObjectBlockEntity
import cc.mirukuneko.realtrainmodrenewed.client.screen.MarkerConfigScreen
import cc.mirukuneko.realtrainmodrenewed.client.screen.ScriptBlockScreen
import cc.mirukuneko.realtrainmodrenewed.client.screen.SignalChangerScreen
import cc.mirukuneko.realtrainmodrenewed.client.screen.SignalReceiverScreen
import cc.mirukuneko.realtrainmodrenewed.client.screen.SignalValueScreen
import cc.mirukuneko.realtrainmodrenewed.client.screen.SpeakerScreen
import cc.mirukuneko.realtrainmodrenewed.client.screen.TrainDetectorScreen
import cc.mirukuneko.realtrainmodrenewed.client.screen.SignalConverterScreen
import cc.mirukuneko.realtrainmodrenewed.client.sound.CrossingGateSoundManager
import cc.mirukuneko.realtrainmodrenewed.installedobject.InstalledObjectCategory
import cc.mirukuneko.realtrainmodrenewed.item.TrainItem
import net.minecraft.client.Minecraft
import net.minecraft.core.BlockPos
import net.minecraft.network.chat.Component
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level

object ClientHooksClient {
    @JvmStatic
    fun openRailSelectScreen(player: Player, stack: ItemStack) {
        ClientItemHelper.openRailSelectScreen(player, stack)
    }

    @JvmStatic
    fun openTrainSelectScreen(player: Player, stack: ItemStack, category: TrainItem.Category) {
        ClientItemHelper.openTrainSelectScreen(player, stack, category)
    }

    @JvmStatic
    fun openTrainSelectScreen(player: Player, stack: ItemStack) {
        ClientItemHelper.openTrainSelectScreen(player, stack)
    }

    @JvmStatic
    fun openVehicleFormationScreen(stack: ItemStack) {
        ClientItemHelper.openVehicleFormationScreen(stack)
    }

    @JvmStatic
    fun openCarSelectScreen(player: Player, stack: ItemStack) {
        ClientItemHelper.openCarSelectScreen(player, stack)
    }

    @JvmStatic
    fun openInstalledObjectSelectScreen(player: Player, stack: ItemStack, category: InstalledObjectCategory) {
        ClientItemHelper.openInstalledObjectSelectScreen(player, stack, category)
    }

    @JvmStatic
    fun openSignalChangerScreen(pos: BlockPos) {
        Minecraft.getInstance().setScreen(SignalChangerScreen(pos))
    }

    @JvmStatic
    fun openSignalReceiverScreen(pos: BlockPos) {
        Minecraft.getInstance().setScreen(SignalReceiverScreen(pos))
    }

    @JvmStatic
    fun openSignalValueScreen(pos: BlockPos) {
        Minecraft.getInstance().setScreen(SignalValueScreen(pos))
    }

    @JvmStatic
    fun openTrainDetectorScreen(pos: BlockPos) {
        Minecraft.getInstance().setScreen(TrainDetectorScreen(pos))
    }

    @JvmStatic
    fun openSignalConverterScreen(pos: BlockPos) {
        Minecraft.getInstance().setScreen(SignalConverterScreen(pos))
    }

    @JvmStatic
    fun openMarkerConfigScreen(pos: BlockPos) {
        Minecraft.getInstance().setScreen(MarkerConfigScreen(pos))
    }

    @JvmStatic
    fun openSpeakerScreen(pos: BlockPos) {
        Minecraft.getInstance().setScreen(SpeakerScreen(pos))
    }

    @JvmStatic
    fun openScriptBlockScreen(pos: BlockPos) {
        Minecraft.getInstance().setScreen(ScriptBlockScreen(pos))
    }

    @JvmStatic
    fun stopCrossingGateSound(level: Level, pos: BlockPos) {
        CrossingGateSoundManager.stop(level, pos)
    }

    @JvmStatic
    fun tickCrossingGateSound(blockEntity: InstalledObjectBlockEntity) {
        CrossingGateSoundManager.tick(blockEntity)
    }

    @JvmStatic
    fun showScriptErrorMessage(message: String?) {
        val minecraft = Minecraft.getInstance()
        val player = minecraft.player
        if (player == null || message.isNullOrBlank()) {
            return
        }
        player.sendSystemMessage(Component.translatable("message.realtrainmodrenewed.script.error", message))
    }
}
